### DISCUSSION TITLE
Sending multiple async requests without waiting on any responses (HTTP Library)

### DISCUSSION THREAD

**Discussion by iSteveZissou** [reactphp/reactphp#529] [Q&A]
As per title, I'm using reactphp http library to send multiple requests as the result of an action - in this context to notify other servers of a change (on this server). I don't need to wait for this response - I simply want to send the requests and move on. So ignoring the response altogether. Is there a way I can do this with the promises - perhaps with timeouts?

The code is working well sending the requests but if one of the servers being notified is slow to return the response I'm blocked waiting on that - so again, I'm looking to completely bypass the response.

Still trying to get my head around async and promises.

Code is a bit like this:

``` php
        foreach ($api_path_array as $connection_record) {
            $browser->post(
                $connection_record['api_path'],
                [
                    'Content-Type' => 'application/json',
                    'Authentication' =>xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
                ],
                json_encode($data)
            )->then(function (\Psr\Http\Message\ResponseInterface $response) use ($connection_record, $data) {
                $msg = "Connection Notification sent to " . $connection_record['api_path'];
                if (isset($data["uid"])) {
                    $msg .= " for uid " . $data["uid"];
                }
                $msg.= ": status " . $response->getStatusCode();
                syslog(LOG_INFO, $msg);
            }, function (\Exception $e) use ($connection_record, $data) {
                $msg = "Error on Callback notification request to " . $connection_record['api_path'];
                if (isset($data["uid"])) {
                    $msg .= " for uid " . $data["uid"];
                }
                $msg .= ": " . $e->getMessage();
                syslog(LOG_DEBUG, $msg);
            });
        }
```

I can somewhat solve this by adding `$browser = $browser->withTimeout(1.0);` to set a custom timeout value - this stops the main request being held up waiting on a response but wondering if there's a more elegant solution here?

Thanks in advance!
Steve

**Comment by SimonFrings** [comment]
Hey @iSteveZissou thanks for bringing this up :+1:

Looking at your code example above, you're currently sending out an HTTP request and waiting for the response before starting to log your messages. In more detail, your code currently sends out a request to the first destination and the [`then()` method](https://github.com/reactphp/promise?tab=readme-ov-file#promiseinterfacethen) defines what to do after the response comes in, depending on if the response was successful or not. The response itself may take a while, so thanks to ReactPHP's async nature, we're not blocking at this point, but instead your code will immediately start a new loop cycle, sending out the next request and so on.

In a gist, you're currently only logging once the response comes in, which obviously requires to wait for the response. If you just want to log directly after sending out the HTTP request, you shouldn't try to log inside the `then()` method, as this function defines what happens after you receive a response.

I hope I understood your problem right and if our conversation answered your questions or helped you progressing, consider supporting the ReactPHP ecosystem by [becoming a sponsor](https://github.com/sponsors/reactphp) ❤️
